guides/10_other-tutorials/building-an-mcp-client-with-gradio.md

```diff
@@ -0,0 +1,489 @@
+        })
+    
+    event_id = response_data["event_id"]
+    
+    # Step 2: Poll for results
+    result_url = f"{base_url}/{event_id}"
+    
+    # Poll with timeout
+    start_time = time.time()
+    timeout = 60
+    
+    while (time.time() - start_time) < timeout:
+        response = session.get(result_url, timeout=15, stream=True)
+        
+        # Process the streaming response
+        for line in response.iter_lines(decode_unicode=True):
+            if line and line.startswith("data:"):
+                try:
+                    data_json = line[5:].strip()
+                    data = json.loads(data_json)
+                    
+                    if isinstance(data, list) and len(data) >= 1:
+                        image_data = data[0]
+                        if isinstance(image_data, dict) and "url" in image_data:
+                            image_url = image_data["url"]
+                            
+                            # Return image info
+                            return json.dumps({
+                                "type": "image",
+                                "url": image_url,
+                                "message": f"Generated image for prompt: {prompt}"
+                            })
+                except:
+                    continue
+        
+        # Wait before polling again
+        time.sleep(2)
+    
+    return json.dumps({
+        "type": "error",
+        "message": "Timeout reached waiting for image generation."
+    })
+
+if __name__ == "__main__":
+    # Initialize and run the server
+    mcp.run(transport='stdio')
+```
+
+### What this server does:
+
+1. It creates an MCP server that exposes a `generate_image` tool
+2. The tool connects to the SanaSprint model hosted on HuggingFace Spaces
+3. It handles the asynchronous nature of image generation by polling for results
+4. When an image is ready, it returns the URL in a structured JSON format
+
+## Part 2: Building the MCP Client with Gradio
+
+Now let's create a Gradio chat interface as MCP Client that connects Claude to our MCP server.
+
+Create a file named `app.py`:
+
+```python
+import asyncio
+import os
+import json
+from typing import List, Dict, Any, Union
+from contextlib import AsyncExitStack
+
+import gradio as gr
+from gradio.components.chatbot import ChatMessage
+from mcp import ClientSession, StdioServerParameters
+from mcp.client.stdio import stdio_client
+from anthropic import Anthropic
+from dotenv import load_dotenv
+
+load_dotenv()  # Load environment variables from .env
+
+# Create event loop for async operations
+loop = asyncio.new_event_loop()
+asyncio.set_event_loop(loop)
+
+class MCPClientWrapper:
+    """Minimal MCP client wrapper for Gradio"""
+    
+    def __init__(self):
+        self.session = None
+        self.exit_stack = None
+        self.anthropic = Anthropic()
+        self.tools = []
+    
+    def connect(self, server_path: str) -> str:
```

Comment:
I think connect can be async if it's just just used in the connect button click event.